feat: Parse and convert units using openAI

Gulianrdgd opened the pull request:
## What type of PR is this?
- feature

## What this PR does / why we need it:
It adds some new routes which allow you to parse the ingredients and convert them using openAI. The code is the same as the parsing but with some additions to the prompt. The results from openAI (at least in my testing) were really great.

## Which issue(s) this PR fixes:
#2731

## Special notes for your reviewer:
I do not have a translation for the button and I am maybe not happy with the place of the button in the UI (maybe the mealie team has some insights). I also do not have a testing function for this, since I am also not very experienced in this field. 

## Testing
I imported many recipes and kept fine-tuning the prompt

Gulianrdgd committed:
Parser sort of works

prompt: I am Dutch so use imperial. You can approximate units like cups and depending on the ingredient use liters or grams. Obviously, ml should be used for ingredients that are liquid and grams for ingredients that are solid. Teaspoons and pinches you can keep the same. You may also convert inches to cm

Gulianrdgd committed:
Made it more general and nicer

Gulianrdgd committed:
Merge branch 'mealie-recipes:mealie-next' into mealie-next

Gulianrdgd committed:
Edited the prompt

Gulianrdgd committed:
Merge remote-tracking branch 'origin/mealie-next' into mealie-next

Gulianrdgd committed:
Remove dev changes

Gulianrdgd committed:
Added temperature to prompt

michael-genson commented:
Would you be able to share some screenshots of the UI? I think that will help others review.
Also, can you share some examples of before/after? Just curious to see some examples (and particularly to share them with users who are bilingual)

Gulianrdgd commented:
Sure! @michael-genson Here are some screenshots

This is the standard recipe just after it has been imported:
![Screenshot 2024-11-04 at 20-38-26 Creamy Macaroni and Cheese](https://github.com/user-attachments/assets/12f115ce-7757-4bc0-9b9f-04d91ef11522)

This is the view of the parser with the original openAI integration:
![Screenshot 2024-11-04 at 20-39-01 Ingredient Parser](https://github.com/user-attachments/assets/3beadde8-db6d-4d8f-bdb0-1c690347f3ec)

This is the view after the conversion:
![Screenshot 2024-11-04 at 20-39-16 Ingredient Parser](https://github.com/user-attachments/assets/d9b263d8-7010-4f3b-a408-3ecb7639a25f)

This is the result:
![Screenshot 2024-11-04 at 20-39-37 Creamy Macaroni and Cheese](https://github.com/user-attachments/assets/91a99176-f1f3-421c-89a5-46c72edd4796)

Gulianrdgd commented:
Examples before and after. The before are withouth any parsing, the second are with the parsing and conversion.

4 tablespoons toasted sesame oil
2 tablespoons soy sauce or tamari
1 tablespoon rice vinegar
1 tablespoon honey
1 (2-inch) piece fresh ginger, peeled and finely grated (about 1 tablespoon)
1 garlic clove, finely grated
1 pound broccoli, trimmed and cut into florets, thick stems discarded
2 scallions, trimmed and cut diagonally into 1 1/2-inch segments, plus thinly sliced scallions for garnish
1 tablespoon olive oil, plus more for brushing the salmon
Kosher salt and black pepper
4 (6-ounce) skin-on salmon fillets
1/2 lime, for serving
Sesame seeds, for serving

=======

60 milliliters sesame oil
30 milliliters soy sauce or tamari
15 milliliters vinegar
15 milliliters honey
1 ginger peeled and finely grated (about 15ml)
1 clove garlic finely grated
450 grams broccoli trimmed and cut into florets, thick stems discarded
2 scallions trimmed and cut diagonally into 4 cm segments, plus thinly sliced scallions for garnish
15 milliliters olive oil plus more for brushing the salmon

680 grams salmon
1/2 lime for serving
sesame seeds for serving

=============

3 tablespoons honey
3/4 teaspoon black pepper, plus more to taste
Kosher salt (such as Diamond Crystal)
2 tablespoons all-purpose flour
1 1/2 teaspoons ground turmeric
1 pound boneless, skinless chicken thighs, cut into 1-inch pieces
1 tablespoon coconut or canola oil
12 ounces asparagus, trimmed and thinly sliced on an angle
1 teaspoon unseasoned rice vinegar or soy sauce
1 lime, cut into wedges (optional)

==========

45 milliliters honey
3/4 teaspoon black pepper plus more to taste
salt such as Diamond Crystal
30 grams flour
1 1/2 teaspoons ground turmeric
450 grams chicken thighs cut into 3cm pieces
15 milliliters coconut oil
340 grams asparagus trimmed and thinly sliced on an angle
1 teaspoon rice vinegar
1 lime cut into wedges (optional)

Gulianrdgd committed:
Merge branch 'mealie-next' into mealie-next

gilbrotheraway commented:
can you please add language translation for recipes using either local whisper or openai API?

khafatech commented:
@Gulianrdgd 

I see some inaccuracies & missing info between the original and the converted recipe.

For example:

* "toasted" is removed in the metrics conversion:
"4 tablespoons toasted sesame oil"
became
"60 milliliters sesame oil"

* The metric version is missing this item also:
"4 (6-ounce) skin-on salmon fillets"

* In the last conversion,  it has non-metrics in the result: "1 1/2 teaspoons ground turmeric"

khafatech commented on docker/Dockerfile:
Is this a mistake, or was there a reason to edit the Dockerfile?

Gulianrdgd commented:
@khafatech 

- I think those differences are there because I removed them when I added the parsed ingredients. For instance, I already has the ingredient sesame oil, so instead of creating a new ingredient, I changed the field ingredient to sesame oil. I also reused most of the prompt which is used in the normal openAI parser, so perhaps these differences are also in that parser.

- I put that into the prompt, Teaspoons and pinches you can keep the same. Because that seemed logical to me. If that isn't the case I could easily change it.

michael-genson commented:
Sorry it took so long to get to this!

In general I like this. Have you considered merging this with the OpenAI parser logic, rather than having it be its own separate prompt? Something like "you may be asked to translate/convert units..." and adding a param. That way we're not duplicating prompts.

I did something similar with the [image parsing prompt](https://github.com/mealie-recipes/mealie/blob/mealie-next/mealie/services/openai/prompts/recipes/parse-recipe-image.txt).

Gulianrdgd commented:
@michael-genson 
No problem! 

I have considered this and I was playing around with it but removed it later. So it's funny that you bring it up.

I had something like you suggested in the beginning but when the prompt got bigger I separated them. I figured that a longer prompt would perform worse., but did not test it.

I'll try to combine the prompts this weekend

michael-genson commented:
Ah okay, if you're getting better results/small requests from keeping it separate that's fine. More of just something to consider. I'll give it a proper look over in the next few days, but overall looks good!

michael-genson commented:
Looks like there are some failing tests on this one

Gulianrdgd committed:
Merge branch 'mealie-recipes:mealie-next' into mealie-next

michael-genson commented:
Hey there! As you might know, we've been working on the upgrade to Nuxt 3 for the past several months, and had to hold on reviewing/merging PRs during that time (especially frontend ones). That work is more-or-less completed, so we've started reviewing PRs again.

This PR has merge conflicts preventing us from fully reviewing/testing it. If you're able to resolve those we can get to work on it.

Appreciate your patience, we know it's been a bit, but now that the Nuxt 3 upgrade is completed we should be able to get things reviewed much quicker than before.

michael-genson commented:
Hey there, this PR has been sitting for a while and has a lot of merge conflicts to be resolved. Since it's been a bit, I'm going to close this PR. Feel free to create a new PR, or you can re-open this one if you're able to resolve the conflicts.
